scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java

```diff
@@ -382,7 +382,10 @@ public void handshakeFailed(Handshaker handshaker, Throwable error) {
 							message.onError(error);
 						}
 					}
-					connectionStore.remove(handshaker.getConnection(), false);
+					Connection connection = handshaker.getConnection();
+					if (!connection.hasEstablishedSession()) {
+						connectionStore.remove(connection, false);
+					}
```

Comment:
In which case, handshake failed but we have an established session ? :thinking:

Reply:
`terminateOngoingHandshake(final Connection connection, final Throwable cause, final AlertDescription description)`

The condition is copied from this function.
The PR is a work-around, I recognized that currently all failing handshakes reports "address change".  A real cleanup of the "alert and termination" requires much more work.

Reply:
It seems you maybe to answer to my question ? 
Do you think or did you face a case where an handshake fails and has an established session ?

It seems impossible to me. :thinking:

Reply:
I guess, the original idea was something as:
- peer 1, currently mapped to endpoint 1 establish successful a session.
- peer 2, then gets mapped to endpoint 1 can't establish successful a session.
- peer 1, then gets mapped to endpoint 1 again.

I don't know, if that works or not. I didn't investigate the "termination".
I just toke the snippet and comment from the other location.

Reply:
> It seems you didn't answer to my question ?

The answer would have required a investigation. Though I didn't investigate, I provided the information, why the code was inserted.

Reply:
As we do not allow `renegociation`, it seems to me that if a handshake failed this implies that there was no session established because handshake is only done if session is not established.

@boaks, If you think I'm right maybe we could keep the code like this but just adding a warning (here and in terminateOngoingHandshake)  in case where handshake failed AND session is established. This way we could see if this happens in real life. WDYT ?

Reply:
OK, go for the warning.

Reply:
You can merge it after that.